overlord/configstate/configcore/corecfg.go

```diff
@@ -62,6 +63,83 @@ func validateBoolFlag(tr config.Conf, flag string) error {
 	return nil
 }
 
+// PlainCoreConfig carries a read-only copy of core config and implements
+// config.ConfReader interface.
+type PlainCoreConfig map[string]interface{}
+
+// Get implements config.ConfReader interface.
+func (cfg PlainCoreConfig) Get(snapName, key string, result interface{}) error {
+	if snapName != "core" {
+		return fmt.Errorf("internal error: expected core snap in Get(), %q was requested", snapName)
+	}
+
+	// XXX: should we support nesting with dotted key?
+	val, ok := cfg[key]
+	if !ok {
+		return &config.NoOptionError{SnapName: snapName, Key: key}
+	}
+
+	rv := reflect.ValueOf(result)
+	rv.Elem().Set(reflect.ValueOf(val))
+	return nil
+}
+
+// Apply applies filesystem modifications under rootDir, according to the
+// cfg configuration. This is a subset of core config options that is important
+// early during boot, before all the configuration is applied as part of
+// normal execution of configure hook.
+func Apply(rootDir string, cfg config.ConfReader) error {
```

Comment:
I'm not super happy that this duplicates some of the code in Run, I think we can avoid that with only a little bit extra complexity, but can go in a follow up.

Reply:
The motivation for this was twofold:
- to avoid check of pending changes (via tr.Changes()) at the start of Run() (this could be avoided by re-organizing the code too).
- to only apply config we are interested in, and avoid e.g. handleProxyConfiguration which needs state (because of assertstate.Store).

Reply:
yes, the goals are fine, we need to see if there's a way to accomplish them and avoid needing to remember to fix two functions when adding new things (which is really my main concern here)